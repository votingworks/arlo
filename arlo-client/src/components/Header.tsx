import React from 'react'
import styled from 'styled-components';


const HeaderContainer = styled.div`
    text-align: center;
    height 100px;
    padding: 20px;
`
const ButtonBar = styled.div `
    display: inline-block;
    float: right;
`

const Button = styled.button`
    background: rgb(211,211,211);
    font-weight: bold;
    font-size: .4em;
    color: black;
    width: 155px;
    height: 30px; 
    border-radius: 5px;
    margin: 0 auto;
`

const Header = () => {

    return (
        <HeaderContainer>
           ARLO
           <ButtonBar>
                <Button>Clear & Restart</Button>
           </ButtonBar>
        </HeaderContainer>
    );

}

export default Header