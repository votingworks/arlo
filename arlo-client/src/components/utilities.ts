import { Params } from '../types'

export const api = <T>(
  endpoint: string,
  { electionId, ...options }: Params & RequestInit
): Promise<T> => {
  const apiBaseURL = electionId ? `/election/${electionId}` : ''
  return fetch(apiBaseURL + endpoint, options).then(res => {
    if (!res.ok) {
      throw new Error(res.statusText)
    }
    return res.json() as Promise<T>
  })
}

export default api
