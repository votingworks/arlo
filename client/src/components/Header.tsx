import React from 'react'
import styled from 'styled-components'
import {
  Navbar,
  NavbarGroup,
  NavbarHeading,
  Alignment,
  NavbarDivider,
} from '@blueprintjs/core'
import { Link, useRouteMatch, RouteComponentProps } from 'react-router-dom'
import { useAuthDataContext } from './UserContext'
import FormButton from './Atoms/Form/FormButton'
import { Inner } from './Atoms/Wrapper'

const ButtonBar = styled.div`
  display: inline-block;
  margin-right: 10px;
`

const Nav = styled(Navbar)`
  width: 100%;
  padding: 0;

  .bp3-navbar-heading img {
    height: 35px;
    padding-top: 5px;
  }

  .bp3-navbar-divider {
    margin-right: 15px;
    margin-left: 0;
  }
`

const InnerBar = styled(Inner)`
  display: inherit;
`

interface TParams {
  electionId: string
  jurisdictionId?: string
}

const Header: React.FC<{}> = () => {
  const electionMatch:
    | RouteComponentProps<TParams>['match']
    | null = useRouteMatch('/election/:electionId')
  const jurisdictionMatch:
    | RouteComponentProps<TParams>['match']
    | null = useRouteMatch(
    '/election/:electionId/jurisdiction/:jurisdictionId?'
  )
  const { isAuthenticated, meta } = useAuthDataContext()
  const electionId = electionMatch ? electionMatch.params.electionId : undefined
  const jurisdiction =
    jurisdictionMatch && meta
      ? meta.jurisdictions.find(
          j => j.id === jurisdictionMatch.params.jurisdictionId
        )
      : undefined
  return (
    <Nav>
      <InnerBar>
        <NavbarGroup align={Alignment.LEFT}>
          <NavbarHeading>
            <Link to="/">
              <img src="/arlo.png" alt="Arlo, by VotingWorks" />
            </Link>
          </NavbarHeading>
          {isAuthenticated && meta!.type === 'jurisdiction_admin' && (
            <NavbarHeading>
              {jurisdiction && `Jurisdiction: ${jurisdiction.name}`}
            </NavbarHeading>
          )}
        </NavbarGroup>
        <NavbarGroup align={Alignment.RIGHT}>
          {isAuthenticated && electionId && meta!.type === 'audit_admin' && (
            <>
              <NavbarHeading>
                <Link to={`/election/${electionId}/setup`}>Audit Setup</Link>
              </NavbarHeading>
              <NavbarDivider />
              <NavbarHeading>
                <Link to={`/election/${electionId}/progress`}>
                  Audit Progress
                </Link>
              </NavbarHeading>
              <NavbarDivider />
              <NavbarHeading>
                <Link to="/">View Audits</Link>
              </NavbarHeading>
              <NavbarDivider />
              <NavbarHeading>
                <Link to="/">New Audit</Link>
              </NavbarHeading>
            </>
          )}
          <ButtonBar id="reset-button-wrapper" />
          {/* istanbul ignore next */
          isAuthenticated && (
            <FormButton
              size="sm"
              onClick={
                /* istanbul ignore next */
                () => window.location.replace('/auth/logout')
              }
            >
              Log out
            </FormButton>
          )}
        </NavbarGroup>
      </InnerBar>
    </Nav>
  )
}

export default Header
