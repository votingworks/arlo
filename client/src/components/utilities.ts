import { useRef, useEffect } from 'react'
import { toast } from 'react-toastify'
import number from '../utils/number-schema'
import { IErrorResponse } from '../types'

const tryJson = (responseText: string) => {
  try {
    return JSON.parse(responseText)
  } catch (err) {
    return {}
  }
}

export const api = async <T>(
  endpoint: string,
  options?: RequestInit
): Promise<T | null> => {
  try {
    const response = await fetch(`/api${endpoint}`, options)
    if (!response.ok) {
      // If we get a 401, it most likely means the session expired, so we
      // redirect to the login screen with a flag to show a message.
      if (response.status === 401) {
        window.location.replace('/#logged-out')
        return null
      }

      const responseText = await response.text()
      const { errors } = tryJson(responseText)
      const error =
        errors && errors.length ? errors[0] : { message: response.statusText }
      const errorData = { ...error, responseText, response }
      console.error(responseText) // eslint-disable-line no-console
      throw errorData
    }
    return response.json() as Promise<T>
  } catch (err) {
    toast.error(err.message)
    return null
  }
}

export const apiDownload = (endpoint: string) => window.open(`/api${endpoint}`)

export const poll = (
  condition: () => Promise<boolean>,
  callback: () => void,
  errback: (arg0: Error) => void,
  timeout: number = 120000,
  interval: number = 1000
) => {
  const endTime = Date.now() + timeout
  ;(async function p() {
    const time = Date.now()
    try {
      const done = await condition()
      if (done) {
        callback()
      } else if (time < endTime) {
        setTimeout(p, interval)
      } else {
        errback(new Error(`Timed out`))
      }
    } catch (err) {
      errback(err)
    }
  })()
}

const numberSchema = number()
  .typeError('Must be a number')
  .integer('Must be an integer')
  .min(0, 'Must be a positive number')
  .required('Required')

export const testNumber = (
  max?: number,
  message?: string
): ((value: number) => Promise<string | undefined>) => {
  const schema = max
    ? numberSchema.concat(
        number().max(max, message || `Must be smaller than ${max}`)
      )
    : numberSchema

  return async (value: unknown) => {
    try {
      await schema.validate(value)
      return undefined
    } catch (error) {
      return error.errors[0]
    }
  }
}

export const asyncForEach = async <T>(
  array: T[],
  callback: (value: T, index: number, array: T[]) => Promise<void>
) => {
  for (let index = 0; index < array.length; index += 1) {
    // eslint-disable-next-line no-await-in-loop
    await callback(array[index], index, array)
  }
}

const getErrorsFromResponse = (
  response: unknown
): { message: string }[] | undefined => {
  if (typeof response !== 'object' || !response) {
    return undefined
  }

  const { errors } = response as { [key: string]: unknown }

  if (!Array.isArray(errors)) {
    return undefined
  }

  return errors
}

export const checkAndToast = (
  response: unknown
): response is IErrorResponse => {
  const errors = getErrorsFromResponse(response)
  if (errors) {
    toast.error(
      `There was a server error regarding: ${errors
        .map(({ message }) => message)
        .join(', ')}`
    )
    return true
  }
  return false
}

// https://overreacted.io/making-setinterval-declarative-with-react-hooks/
/* istanbul ignore next */
export const useInterval = (
  callback: Function,
  delay: number | null,
  callImmediately?: boolean
) => {
  const savedCallback = useRef<Function>()

  // Remember the latest function.
  useEffect(() => {
    savedCallback.current = callback
  }, [callback])

  // Set up the interval.
  useEffect(/* eslint consistent-return: "off" */
  () => {
    function tick() {
      savedCallback.current!()
    }
    if (callImmediately) tick()
    if (delay !== null) {
      // allows for pausing
      const id = setInterval(tick, delay)
      return () => clearInterval(id)
    }
  }, [delay, callImmediately])
}

/** https://stackoverflow.com/questions/6229197/how-to-know-if-two-arrays-have-the-same-values/55614659#55614659
 * assumes array elements are primitive types
 * check whether 2 arrays are equal sets.
 * @param  {} a1 is an array
 * @param  {} a2 is an array
 */
/* istanbul ignore next */
export const areArraysEqualSets = (a1: unknown[], a2: unknown[]): boolean => {
  const superSet: { [key: string]: number } = {}
  for (const i of a1) {
    const e = i + typeof i
    superSet[e] = 1
  }

  for (const i of a2) {
    const e = i + typeof i
    if (!superSet[e]) {
      return false
    }
    superSet[e] = 2
  }

  for (const e in superSet) {
    if (superSet[e] === 1) {
      return false
    }
  }

  return true
}
