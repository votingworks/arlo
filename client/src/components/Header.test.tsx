import React from 'react'
import { render, waitFor } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'
import Header from './Header'
import * as utilities from './utilities'
import AuthDataProvider from './UserContext'

const apiMock: jest.SpyInstance<
  ReturnType<typeof utilities.api>,
  Parameters<typeof utilities.api>
> = jest.spyOn(utilities, 'api').mockImplementation()
const checkAndToastMock: jest.SpyInstance<
  ReturnType<typeof utilities.checkAndToast>,
  Parameters<typeof utilities.checkAndToast>
> = jest.spyOn(utilities, 'checkAndToast').mockReturnValue(false)

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useRouteMatch: () => ({
    url: '/election/1',
    params: {
      electionId: '1',
    },
  }),
}))

checkAndToastMock.mockReturnValue(false)

afterEach(() => {
  apiMock.mockClear()
  checkAndToastMock.mockClear()
})

describe('Header', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Router>
        <Header />
      </Router>
    )
    expect(container).toMatchSnapshot()
  })

  it('shows the logout button when authenticated', async () => {
    apiMock.mockImplementation(async () => ({
      type: 'audit_admin',
      name: 'Joe',
      email: 'test@email.org',
      jurisdictions: [],
      organizations: [],
    }))
    const { findByText } = render(
      <Router>
        <AuthDataProvider>
          <Header />
        </AuthDataProvider>
      </Router>
    )

    await findByText('Log out')
    expect(apiMock).toHaveBeenCalledTimes(1)
    expect(apiMock).toHaveBeenCalledWith('/me')
  })

  it('does not show logout button if not authenticated', async () => {
    apiMock.mockRejectedValue(async () => ({}))
    const { queryByText } = render(
      <Router>
        <AuthDataProvider>
          <Header />
        </AuthDataProvider>
      </Router>
    )

    const loginButton = queryByText('Log out')
    await waitFor(() => {
      expect(apiMock).toHaveBeenCalledTimes(1)
      expect(apiMock).toHaveBeenCalledWith('/me')
      expect(loginButton).toBeFalsy()
    })
  })

  it('does not show logout button if the authentication verification has a server error', async () => {
    checkAndToastMock.mockReturnValue(true)
    apiMock.mockRejectedValue(async () => ({}))
    const { queryByText } = render(
      <Router>
        <AuthDataProvider>
          <Header />
        </AuthDataProvider>
      </Router>
    )

    const loginButton = queryByText('Log out')
    await waitFor(() => {
      expect(apiMock).toHaveBeenCalledTimes(1)
      expect(apiMock).toHaveBeenCalledWith('/me')
      expect(loginButton).toBeFalsy()
    })
  })

  it('shows the nav bar when authenticated and there is an electionId', async () => {
    checkAndToastMock.mockReturnValue(false)
    apiMock.mockImplementation(async () => ({
      type: 'audit_admin',
      name: 'Joe',
      email: 'test@email.org',
      jurisdictions: [],
      organizations: [],
    }))
    const { container, getByText } = render(
      <Router>
        <AuthDataProvider>
          <Header />
        </AuthDataProvider>
      </Router>
    )
    await waitFor(() => {
      expect(getByText('Audit Setup')).toBeTruthy()
      expect(getByText('Audit Progress')).toBeTruthy()
      expect(getByText('View Audits')).toBeTruthy()
      expect(getByText('New Audit')).toBeTruthy()
      expect(container).toMatchSnapshot()
    })
  })
})
