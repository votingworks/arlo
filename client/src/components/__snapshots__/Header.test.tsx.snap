// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<div>
  <div
    class="bp3-navbar sc-EHOje kGnsSl"
  >
    <div
      class="sc-bxivhb sc-bZQynM ftKVqf"
    >
      <div
        class="bp3-navbar-group bp3-align-left"
      >
        <div
          class="bp3-navbar-heading"
        >
          <a
            href="/"
          >
            <img
              alt="Arlo, by VotingWorks"
              src="/arlo.png"
            />
          </a>
        </div>
      </div>
      <div
        class="bp3-navbar-group bp3-align-right"
      >
        <div
          class="sc-ifAKCX jNLPyG"
          id="reset-button-wrapper"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Header shows the nav bar when authenticated and there is an electionId 1`] = `
<div>
  <div
    class="bp3-navbar sc-EHOje kGnsSl"
  >
    <div
      class="sc-bxivhb sc-bZQynM ftKVqf"
    >
      <div
        class="bp3-navbar-group bp3-align-left"
      >
        <div
          class="bp3-navbar-heading"
        >
          <a
            href="/"
          >
            <img
              alt="Arlo, by VotingWorks"
              src="/arlo.png"
            />
          </a>
        </div>
      </div>
      <div
        class="bp3-navbar-group bp3-align-right"
      >
        <div
          class="bp3-navbar-heading"
        >
          <a
            href="/election/1/setup"
          >
            Audit Setup
          </a>
        </div>
        <div
          class="bp3-navbar-divider"
        />
        <div
          class="bp3-navbar-heading"
        >
          <a
            href="/election/1/progress"
          >
            Audit Progress
          </a>
        </div>
        <div
          class="bp3-navbar-divider"
        />
        <div
          class="bp3-navbar-heading"
        >
          <a
            href="/"
          >
            View Audits
          </a>
        </div>
        <div
          class="bp3-navbar-divider"
        />
        <div
          class="bp3-navbar-heading"
        >
          <a
            href="/"
          >
            New Audit
          </a>
        </div>
        <div
          class="sc-ifAKCX jNLPyG"
          id="reset-button-wrapper"
        />
        <button
          class="bp3-button bp3-small sc-bwzfXH eBfcXh"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            Log out
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
